std/parallelism.d

```diff
@@ -4303,243 +4321,341 @@ version(parallelismStressTest)
-            {
-                stderr.writeln(
-                    "Success doing matrix stuff that involves nested pool use.");
-            });
-            poolInstance.put(saySuccess);
-            saySuccess.workForce();
+    // Test worker-local storage.
+    auto workerLocalStorage = poolInstance.workerLocalStorage!uint(1);
+    foreach (i; poolInstance.parallel(iota(0U, 1_000_000)))
+    {
+        workerLocalStorage.get++;
+    }
+    assert(reduce!"a + b"(workerLocalStorage.toRange) ==
+    1_000_000 + poolInstance.size + 1);
 
-            // A more thorough test of amap, reduce:  Find the sum of the square roots of
-            // matrix.
+    // Make sure work is reasonably balanced among threads.  This test is
+    // non-deterministic and is more of a sanity check than something that
+    // has an absolute pass/fail.
+    shared(uint)[void*] nJobsByThread;
+    foreach (thread; poolInstance.pool)
+    {
+        nJobsByThread[cast(void*) thread] = 0;
+    }
+    nJobsByThread[ cast(void*) Thread.getThis()] = 0;
 
-            static real parallelSum(real[] input)
-            {
-                return poolInstance.reduce!"a + b"(input);
-            }
+    foreach (i; poolInstance.parallel( iota(0, 1_000_000), 100 ))
+    {
+        atomicOp!"+="( nJobsByThread[ cast(void*) Thread.getThis() ], 1);
+    }
 
-            auto sumSqrt = poolInstance.reduce!"a + b"(
-                               poolInstance.amap!parallelSum(
-                                   sqrtMatrix
-                               )
-                           );
+    version(parallelismStressTest) stderr.writeln("\nCurrent thread is:  ",
+    cast(void*) Thread.getThis());
+    version(parallelismStressTest) stderr.writeln("Workload distribution:  ");
 
-            assert(approxEqual(sumSqrt, 4.437e8));
-            stderr.writeln("Done sum of square roots.");
+    version(parallelismStressTest)
+    {
+        foreach (k, v; nJobsByThread)
+        {
+            stderr.writeln(k, '\t', v);
+        }
+    }
+    else
+    {
+        foreach (k, v; nJobsByThread)
+        {
+        }
+    }
 
-            // Test whether tasks work with function pointers.
-            auto nanTask = task(&isNaN, 1.0L);
-            poolInstance.put(nanTask);
-            assert(nanTask.spinForce == false);
+    // Test whether amap can be nested.
+    real[][] matrix = new real[][](1000, 1000);
+    foreach (i; poolInstance.parallel( iota(0, matrix.length) ))
+    {
+        foreach (j; poolInstance.parallel( iota(0, matrix[0].length) ))
+        {
+            matrix[i][j] = i * j;
+        }
+    }
 
-            if (poolInstance.size > 0)
-            {
-                // Test work waiting.
-                static void uselessFun()
-                {
-                    foreach (i; 0..1_000_000) {}
-                }
+    // Get around weird bugs having to do w/ sqrt being an intrinsic:
+    //real[][] sqrtMatrix = poolInstance.amap!parallelSqrt(matrix);
```

Comment:
amap can not call parallelSqrt that that local function was trying to access poolInstance which is located on the same stack